packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,126 @@
+
+    async getOrCreateWidget(options: TaskTerminalWidgetOpenerOptions): Promise<TerminalWidget> {
+        const isDedicated = TaskTerminalWidgetOpenerOptions.isDedicatedTerminal(options);
+        if (isDedicated && !options.taskConfig) {
+            throw new Error('"taskConfig" must be included as part of the "option" if "isDedicated" is true');
+        }
+
+        const { isNew, widget } = await this.getWidgetToRunTask(options);
+        if (isNew) {
+            this.shell.addWidget(widget, { area: options.widgetOptions ? options.widgetOptions.area : 'bottom' });
+            this.terminalWidgetMap.set(widget.id, {
+                isDedicated, taskConfig: options.taskConfig, widget
+            });
+        } else if (options.title) {
+            widget.setTitle(options.title);
+        }
+
+        this.terminalService.open(widget, options);
+
+        return widget;
+    }
+
+    protected async getWidgetToRunTask(options: TaskTerminalWidgetOpenerOptions): Promise<{ isNew: boolean, widget: TerminalWidget }> {
+        let reusableTerminalWidget: TerminalWidget | undefined;
+        if (TaskTerminalWidgetOpenerOptions.isDedicatedTerminal(options)) {
+            for (const { isDedicated, taskConfig, widget } of this.terminalWidgetMap.values()) {
+                // to run a task whose `taskPresentation === 'dedicated'`, the terminal to be reused must be
+                // 1) dedicated, 2) idle, and 3) the one that ran the same task
+                if (isDedicated &&
+                    !widget.hasRunningTask &&
+                    taskConfig &&
+                    options.taskConfig &&
+                    this.taskDefinitionRegistry.compareTasks(options.taskConfig, taskConfig)) {
+
+                    reusableTerminalWidget = widget;
+                    break;
+                }
+            }
+        } else if (TaskTerminalWidgetOpenerOptions.isSharedTerminal(options)) {
+            const availableWidgets: TerminalWidget[] = [];
+            for (const { isDedicated, widget } of this.terminalWidgetMap.values()) {
+                // to run a task whose `taskPresentation === 'shared'`, the terminal to be used must be
+                // 1) not dedicated, and 2) idle
+                if (!isDedicated && !widget.hasRunningTask) {
+                    availableWidgets.push(widget);
+                }
+            }
+            const currentWidget = availableWidgets.find(w => {
+                const currentTerminal = this.terminalService.currentTerminal;
+                return currentTerminal && currentTerminal.id === w.id;
+            });
+            reusableTerminalWidget = currentWidget || availableWidgets[0];
+        }
+
+        // we are unable to find a terminal widget to run the task, or `taskPresentation === 'new'`
+        if (!reusableTerminalWidget) {
+            const widget = <TerminalWidget>await this.widgetManager.getOrCreateWidget(TERMINAL_WIDGET_FACTORY_ID, options);
```

Comment:
`TERMINAL_WIDGET_FACTORY_ID` should implementation details of `TerminalService`, maybe you are better to use `TerminalService.newTerminal` and get rid of dependencies to `WidgetManager`